src/Microsoft.Data.SqlClient/netcore/src/Common/src/System/Net/Security/NegotiateStreamPal.Unix.cs

```diff
@@ -156,20 +159,33 @@ internal static partial class NegotiateStreamPal
             }
 
             SafeFreeNegoCredentials negoCredentialsHandle = (SafeFreeNegoCredentials)credentialsHandle;
+            SecurityStatusPal status = default;
 
-            if (negoCredentialsHandle.IsDefault && string.IsNullOrEmpty(spn))
+            foreach (string spn in spns)
             {
-                throw new PlatformNotSupportedException(Strings.net_nego_not_supported_empty_target_with_defaultcreds);
-            }
+                if (negoCredentialsHandle.IsDefault && string.IsNullOrEmpty(spn))
+                {
+                    throw new PlatformNotSupportedException(Strings.net_nego_not_supported_empty_target_with_defaultcreds);
+                }
 
-            SecurityStatusPal status = EstablishSecurityContext(
-                negoCredentialsHandle,
-                ref securityContext,
-                spn,
-                requestedContextFlags,
-                ((inSecurityBufferArray != null && inSecurityBufferArray.Length != 0) ? inSecurityBufferArray[0] : null),
-                outSecurityBuffer,
-                ref contextFlags);
+                status = EstablishSecurityContext(
+                    negoCredentialsHandle,
+                    ref securityContext,
+                    spn,
+                    requestedContextFlags,
+                    ((inSecurityBufferArray != null && inSecurityBufferArray.Length != 0) ? inSecurityBufferArray[0] : null),
+                    outSecurityBuffer,
+                    ref contextFlags);
+
+                if (status.ErrorCode != SecurityStatusPalErrorCode.InternalError)
+                {
+                    break; // Successful case, exit the loop with current SPN.
+                }
+                else
+                {
+                    securityContext = null; // Reset security context to be generated again for next SPN.
```

Comment:
`securityContext` will be set to null even if we are not going to retry with a different spn if `SecurityStatusPalErrorCode.InternalError` is returned from `EstablishSecurityContext()`. That null will be returned back to the caller. This is kind of changing the initial behavior when we have only 1 SPN which fails with InternalError. 
Can we only set `securityContext = null` when we need an additional try in the for loop?

Reply:
I thought about it - this is an error event - we eitherway don't need securityContext anymore.
We're going to throw exception in that case.